tracer/slog.go

```diff
@@ -0,0 +1,109 @@
+package tracer
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"log/slog"
+	"strings"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+)
+
+type SlogTracer struct {
+	Logger           *slog.Logger
+	LogAllQueries    bool
+	LogFailedQueries bool
+	// log passed params
+	// replace placeholders with arguments useful for local debugging
+	LogValues bool
+	// enabled if non-zero value is provided
+	LogSlowQueriesThreshold time.Duration
+}
+
+func NewSlogTracer(logger *slog.Logger, opts ...Option) *SlogTracer {
+	cfg := &config{
+		logAllQueries:           false,
+		logFailedQueries:        false,
+		logValues:               false,
+		logSlowQueriesThreshold: 0,
+	}
+
+	for _, opt := range opts {
+		opt.apply(cfg)
+	}
+
+	return &SlogTracer{
+		Logger:                  logger,
+		LogAllQueries:           cfg.logAllQueries,
+		LogFailedQueries:        cfg.logFailedQueries,
+		LogValues:               cfg.logValues,
+		LogSlowQueriesThreshold: cfg.logSlowQueriesThreshold,
+	}
+}
+
+func (s *SlogTracer) TraceQueryStart(ctx context.Context, _ *pgx.Conn, data pgx.TraceQueryStartData) context.Context {
+	query := data.SQL
+	if s.LogValues {
+		for i, placeholder := range data.Args {
+			query = strings.Replace(query, fmt.Sprintf("$%d", i+1), fmt.Sprintf("%v", placeholder), 1)
+		}
+	}
+
+	if s.LogAllQueries || isTracingEnabled(ctx) {
+		if s.LogValues {
+			s.Logger.Info("query start", slog.String("query", query), slog.Any("args", data.Args))
```

Comment:
This is little bit faster.
```suggestion
			s.Logger.LogAttrs(ctx, slog.LevelInfo, "query start", slog.String("query", query), slog.Any("args", data.Args))
```
But i would be nice, if the user would be able to set his own levels for info, warn and error, so like this
```suggestion
			s.Logger.LogAttrs(ctx, s.LevelInfo, "query start", slog.String("query", query), slog.Any("args", data.Args))
```